tmt/steps/report/polarion.py

```diff
@@ -213,93 +224,119 @@ def go(self, *, logger: Optional[tmt.log.Logger] = None) -> None:
-            'polarion-project-span-ids': project_id}
         for tr_field in other_testrun_fields:
             param = self.get(tr_field, os.getenv(f'TMT_PLUGIN_REPORT_POLARION_{tr_field.upper()}'))
             # TODO: remove the os.getenv when envvars in click work with steps in plans as well
             # as with steps on cmdline
             if param:
-                properties[f"polarion-custom-{tr_field.replace('_', '')}"] = param
+                testsuites_properties[f"polarion-custom-{tr_field.replace('_', '')}"] = param
+
         if use_facts:
-            properties['polarion-custom-hostname'] = \
-                self.step.plan.provision.guests()[0].primary_address
-            properties['polarion-custom-arch'] = self.step.plan.provision.guests()[0].facts.arch
+            guests = self.step.plan.provision.guests()
+            try:
+                testsuites_properties['polarion-custom-hostname'] = guests[0].primary_address
+                testsuites_properties['polarion-custom-arch'] = guests[0].facts.arch
+            except IndexError as e:
+                raise tmt.utils.ReportError(
+                    f'Failed to get the Polarion facts from `step.plan.provision.guests()`: {e}')
+
         if template:
-            properties['polarion-testrun-template-id'] = template
+            testsuites_properties['polarion-testrun-template-id'] = template
+
         logs = os.getenv('TMT_REPORT_ARTIFACTS_URL')
-        if logs and 'polarion-custom-logs' not in properties:
-            properties['polarion-custom-logs'] = logs
-        testsuites_properties = ElementTree.SubElement(xml_tree, 'properties')
-        for name, value in properties.items():
-            ElementTree.SubElement(testsuites_properties, 'property', attrib={
-                'name': name, 'value': str(value)})
-
-        testsuite = xml_tree.find('testsuite')
-        project_span_ids = xml_tree.find(
-            '*property[@name="polarion-project-span-ids"]')
-
-        for result in self.step.plan.execute.results():
+        if logs and 'polarion-custom-logs' not in testsuites_properties:
+            testsuites_properties['polarion-custom-logs'] = logs
+
+        project_span_ids: list[str] = []
+
+        results_context = ResultsContext(self.step.plan.execute.results())
+
+        for result in results_context:
             if not result.ids or not any(result.ids.values()):
                 self.warn(
                     f"Test Case '{result.name}' is not exported to Polarion, "
                     "please run 'tmt tests export --how polarion' on it.")
                 continue
+
             work_item_id, test_project_id = find_polarion_case_ids(result.ids)
 
             if test_project_id is None:
                 self.warn(f"Test case '{result.name}' missing or not found in Polarion.")
                 continue
 
+            # TODO: try to remove the asserts by raising a proper ReportError exception?
```

Comment:
@KwisatzHaderach 
Same here - What about checking `work_item_id` and `project_span_id` for `None` value and raising a proper exception instead of failing completely without a proper error message for the user?